dns_sd_avahi.c

```diff
@@ -125,18 +362,47 @@ int discover_host(char *addr_str, size_t addr_len, uint16_t *port)
 	}
 
 	DEBUG("Trying to discover host\n");
-	avahi_simple_poll_loop(poll);
+	avahi_simple_poll_loop(d->poll);
 
-	if (!ddata.found)
+	if (!d->resolved)
 		ret = ENXIO;
 
-	if (ret == 0)
-		avahi_address_snprint(addr_str, addr_len, &address);
+	if (ret == 0) {
+		ret = port_knock_discovery_data(&d);
+		remove_dup_discovery_data(&d);
+		t = d;
+		while (t->next) {
+			t = t->next;
+			ret++;
+		}
+	}
 
 	avahi_service_browser_free(browser);
 err_free_client:
 	avahi_client_free(client);
 err_free_poll:
-	avahi_simple_poll_free(poll);
-	return -ret; /* we want a negative error code */
+	avahi_simple_poll_free(d->poll);
+	iio_mutex_destroy(d->lock);
+	*ddata = d;
+
+	return ret;
+}
+
+int discover_host(char *addr_str, size_t addr_len, uint16_t *port)
+{
+	struct dns_sd_discovery_data *ddata;
+	int ret = 0;
+
+	ret = dnssd_find_hosts(&ddata);
+
```

Comment:
I propose that this function fails if discovery finds more than one IIOD server. Behaviour is not deterministic otherwise, because it is no guaranteed that the first element on the list is always the same.

Reply:
correct - the proposed solution is that the first that the DNS SD passes back is the one that will get passed to the user.

This is what happens today as well - it's just we don't decode things. In today's code - the last one which goes into the single buffer is the one that is passed back...

Reply:
Yes, this change would break backward compatibility. Is it possible to somehow deprecate the functionality and remove it in one or two releases?

Reply:
It's sort of up to the application that uses libiio if they want to pass blank ip: context names to the library or not. As long as people understand what the library does, it should be OK.(maybe?)

What I will do in things like iio_attr is use the rest of the arguments that are passed to the application as "hints" to pass back the right context, and only fail if there are multiple contexts that match. Same as what the hints argument does for `getaddrinfo()`.

Things like:
```
rgetz@brain:~/github/libiio$ iio_attr -a -C hw_model
Using auto-detected IIO context at URI "usb:3.15.5"
hw_model: Analog Devices PlutoSDR Rev.B (Z7010-AD9364)
```

will only work if no other context has a context attribute of hw_model. 
Right now, it does:
```
rgetz@brain:~/github/libiio-dns/build$ ./tests/iio_attr -a -C hw_model
Multiple contexts found. Please select one using --uri:
	0: 0456:b673 (Analog Devices Inc. PlutoSDR (ADALM-PLUTO)), serial=104473222a87000618000600473ed57ae0 [usb:3.15.5]
	1: 192.168.1.120 (ad7291,ad9361-phy,xadc,cf-ad9361-dds-core-lpc,cf-ad9361-lpc) [ip:analog.local]
```

which is silly, since only one context has a context attribute with that name.